gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcClientCalls.java

```diff
@@ -90,4 +99,101 @@
+   * wait for trailers to resolve a unary RPC. This can save milliseconds when the server is
+   * overloaded.
+   */
+  @BetaApi
+  static <RequestT, ResponseT> ApiFuture<ResponseT> eagerFutureUnaryCall(
+      ClientCall<RequestT, ResponseT> clientCall, RequestT request) {
+    // Start the call
+    GrpcFuture<ResponseT> future = new GrpcFuture<>(clientCall);
+    clientCall.start(new EagerFutureListener<>(future), new Metadata());
+
+    // Send the request
+    try {
+      clientCall.sendMessage(request);
+      clientCall.halfClose();
+      // Request an extra message to detect misconfigured servers
+      clientCall.request(2);
+    } catch (Throwable sendError) {
+      // Cancel if anything goes wrong and
+      try {
+        clientCall.cancel(null, sendError);
+      } catch (Throwable cancelError) {
+        LOGGER.log(Level.SEVERE, "Error encountered while closing it", sendError);
+      }
+
+      throw sendError;
+    }
+
+    return future;
+  }
+
+  /** Thin wrapper around an ApiFuture that will cancel the underlying ClientCall. */
+  private static class GrpcFuture<T> extends AbstractApiFuture<T> {
+    private final ClientCall<?, T> call;
+
+    private GrpcFuture(ClientCall<?, T> call) {
+      this.call = call;
+    }
+
+    @Override
+    protected void interruptTask() {
+      call.cancel("GrpcFuture was cancelled", null);
+    }
+
+    @Override
+    public boolean set(T value) {
+      return super.set(value);
+    }
+
+    @Override
+    public boolean setException(Throwable throwable) {
+      return super.setException(throwable);
+    }
+  }
+
+  /**
+   * A bridge between gRPC's ClientCall.Listener to an ApiFuture.
+   *
+   * <p>The Listener will eagerly resolve the future when the first message is received and will not
+   * wait for the trailers. This should cut down on the latency at the expense of safety. If the
+   * server is misconfigured and sends a second response for a unary call, the error will be logged,
+   * but the future will still be successful.
+   */
+  private static class EagerFutureListener<T> extends ClientCall.Listener<T> {
+    private final GrpcFuture<T> future;
+
+    private EagerFutureListener(GrpcFuture<T> future) {
+      this.future = future;
+    }
+
+    @Override
+    public void onMessage(T message) {
+      if (!future.set(message)) {
+        throw Status.INTERNAL
+            .withDescription("More than one value received for unary call")
+            .asRuntimeException();
+      }
+    }
+
+    @Override
+    public void onClose(Status status, Metadata trailers) {
+      if (!future.isDone()) {
+        future.setException(
+            Status.INTERNAL
+                .withDescription("No value received for unary call")
+                .asException(trailers));
+      }
+      if (!status.isOk()) {
+        LOGGER.log(
+            Level.SEVERE, "Received error for unary call after receiving a successful response");
```

Comment:
Similarly to the previous comment, I don't think it should be SEVERE level here. In general, we do not log anything on this level (by level I mean not loggin level, but architectural level of a raw GrpcCall), instead an exception is thrown, which will get logged eventually on a higher leverl. I.e. instead of trying to log the low-level grpc failures here, ideally we should make sure that the proper exception gets propagated to the higher, more user-visible level, where it can be either handled properly or logged at least.

Reply:
Unfortunately we cant propagate the exception here because the future has already been resolved. I downgraded the level to WARNING